core/vm/lightclient/lightclient.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	encodingBytes := make([]byte, serializeLength)
+
+	pos := uint64(0)
+	if uint64(len(cs.ChainID)) > chainIDLength {
+		return nil, fmt.Errorf("chainID length should be no more than 32")
+	}
+	copy(encodingBytes[pos:pos+chainIDLength], cs.ChainID)
+	pos += chainIDLength
+
+	binary.BigEndian.PutUint64(encodingBytes[pos:pos+heightLength], cs.Height)
+	pos += heightLength
+
+	copy(encodingBytes[pos:pos+validatorSetHashLength], cs.NextValidatorSetHash)
+	pos += validatorSetHashLength
+
+	for index := uint64(0); index < validatorSetLength; index++ {
+		validator := cs.ValidatorSet.Validators[index]
+		pubkey, ok := validator.PubKey.(ed25519.PubKey)
+		if !ok {
+			return nil, fmt.Errorf("invalid pubkey type")
+		}
+
+		copy(encodingBytes[pos:pos+validatorPubkeyLength], pubkey[:])
+		pos += validatorPubkeyLength
+
+		binary.BigEndian.PutUint64(encodingBytes[pos:pos+validatorVotingPowerLength], uint64(validator.VotingPower))
+		pos += validatorVotingPowerLength
+
+		copy(encodingBytes[pos:pos+relayerAddressLength], validator.RelayerAddress)
+		pos += relayerAddressLength
+
+		copy(encodingBytes[pos:pos+relayerBlsKeyLength], validator.BlsKey)
+		pos += relayerBlsKeyLength
+	}
+
+	return encodingBytes, nil
+}
+
+func (cs *ConsensusState) ApplyLightBlock(block *types.LightBlock) (bool, error) {
+	if uint64(block.Height) <= cs.Height {
+		return false, fmt.Errorf("block height <= consensus height (%d < %d)", block.Height, cs.Height)
+	}
+
+	if err := block.ValidateBasic(cs.ChainID); err != nil {
+		return false, err
+	}
+
+	if cs.Height == uint64(block.Height-1) {
+		if !bytes.Equal(cs.NextValidatorSetHash, block.ValidatorsHash) {
+			return false, fmt.Errorf("validators hash mismatch, expected: %s, real: %s", cs.NextValidatorSetHash, block.ValidatorsHash)
+		}
+		err := block.ValidatorSet.VerifyCommitLight(cs.ChainID, block.Commit.BlockID, block.Height, block.Commit)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		// Ensure that +`trustLevel` (default 1/3) or more of last trusted validators signed correctly.
+		err := cs.ValidatorSet.VerifyCommitLightTrusting(cs.ChainID, block.Commit, light.DefaultTrustLevel)
+		if err != nil {
+			return false, err
+		}
+
+		// Ensure that +2/3 of new validators signed correctly.
+		//
+		// NOTE: this should always be the last check because untrustedVals can be
+		// intentionally made very large to DOS the light client. not the case for
+		// VerifyAdjacent, where validator set is known in advance.
+		err = block.ValidatorSet.VerifyCommitLight(cs.ChainID, block.Commit.BlockID, block.Height, block.Commit)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// update consensus state
+	cs.Height = uint64(block.Height)
+	cs.NextValidatorSetHash = block.NextValidatorsHash
+	cs.ValidatorSet = block.ValidatorSet
+
+	return !(bytes.Equal(cs.ValidatorSet.Hash(), block.ValidatorsHash)), nil
```

Comment:
This is not correct, please check the newest implementation, there is a bug here.
https://github.com/bnb-chain/bsc/blob/master/core/vm/lightclient/v2/lightclient.go#L121